File: app/src/main/java/awais/instagrabber/models/Resource.java
package awais.instagrabber.models;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class Resource<T> {
    public final Status status;
    public final T data;
    public final String message;

    private Resource(@NonNull Status status,
                     @Nullable T data,
                     @Nullable String message) {
        this.status = status;
        this.data = data;
        this.message = message;
    }

    @NonNull
    public static <T> Resource<T> success(@NonNull T data) {
        return new Resource<>(Status.SUCCESS, data, null);
    }

    @NonNull
    public static <T> Resource<T> error(String msg, @Nullable T data) {
        return new Resource<>(Status.ERROR, data, msg);
    }

    @NonNull
    public static <T> Resource<T> loading(@Nullable T data) {
        return new Resource<>(Status.LOADING, data, null);
    }

    public enum Status {
        SUCCESS,
        ERROR,
        LOADING
    }
}

File: app/src/main/java/awais/instagrabber/models/Resource.kt
package awais.instagrabber.models

import androidx.annotation.StringRes

data class Resource<T>(
    @JvmField val status: Status,
    @JvmField val data: T? = null,
    @JvmField val message: String? = null,
    @JvmField @StringRes val resId: Int = 0,
) {
    enum class Status {
        SUCCESS, ERROR, LOADING
    }

    companion object {
        @JvmStatic
        fun <T> success(data: T): Resource<T> {
            return Resource(Status.SUCCESS, data, null, 0)
        }

        @JvmStatic
        fun <T> error(msg: String?, data: T?): Resource<T?> {
            return Resource(Status.ERROR, data, msg, 0)
        }

        @JvmStatic
        fun <T> error(resId: Int, data: T?): Resource<T?> {
            return Resource(Status.ERROR, data, null, resId)
        }

        @JvmStatic
        fun <T> loading(data: T?): Resource<T?> {
            return Resource(Status.LOADING, data, null, 0)
        }
    }
}